[WebGPU] GPUAutoLayout should have special designation

mwyrzykowski opened the pull request:
#### fc7310ff1cabbb2f44a8259c1fca73f93958396d
<pre>
[WebGPU] GPUAutoLayout should have special designation
<a href="https://bugs.webkit.org/show_bug.cgi?id=257947">https://bugs.webkit.org/show_bug.cgi?id=257947</a>
&lt;radar://110624583&gt;

Reviewed by Myles C. Maxfield.

Instead of assuming zero bind group layouts == auto layout, use
an optional to track auto layout.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::m_autoPipelineLayout):
(WebCore::GPUDevice::createAutoPipelineLayout):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createPipelineLayout):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUPipelineLayoutDescriptor.h:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/PipelineLayout.h:
(WebGPU::PipelineLayout::create):
(WebGPU::PipelineLayout::isValid const):
(WebGPU::PipelineLayout::isAutoLayout const):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(WebGPU::Device::createPipelineLayout):
(WebGPU::PipelineLayout::PipelineLayout):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUPipelineLayoutDescriptor.h:

Canonical link: <a href="https://commits.webkit.org/265134@main">https://commits.webkit.org/265134@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c2fe834ab5ff7541e6867187c6acb055c351675

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9638 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10146 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12580 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10102 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10892 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8423 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11982 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8199 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9017 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16356 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9298 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9167 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12440 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9670 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8829 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2381 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13056 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9445 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebKit/Shared/WebGPU/WebGPUPipelineLayoutDescriptor.serialization.in:
```diff
@@ -23,5 +23,6 @@
 #if ENABLE(GPU_PROCESS)
 [AdditionalEncoder=StreamConnectionEncoder] struct WebKit::WebGPU::PipelineLayoutDescriptor : WebKit::WebGPU::ObjectDescriptorBase {
     Vector<WebKit::WebGPUIdentifier> bindGroupLayouts;
+    bool isAutoLayout;
```
Shouldn't this be represented in-line as an optional?

mwyrzykowski replied:
oh make `bindGroupLayouts` an optional?

litherum commented on Source/WebKit/Shared/WebGPU/WebGPUPipelineLayoutDescriptor.h:
```diff
@@ -36,6 +36,7 @@ namespace WebKit::WebGPU {
 
 struct PipelineLayoutDescriptor : public ObjectDescriptorBase {
     Vector<WebGPUIdentifier> bindGroupLayouts;
+    bool isAutoLayout { false };
```
Ditto.

litherum reviewed (approved):
Please consider using optionals wherever it makes sense.

tadeuzagallo commented on Source/WebGPU/WebGPU/ComputePipeline.mm:
```diff
@@ -88,7 +88,7 @@ static MTLSize metalSize(auto workgroupSize)
 
     // FIXME: https://bugs.webkit.org/show_bug.cgi?id=257947 - GPUAutoLayout should not rely on the only
```
nit: I think this fixes this FIXME, right?

tadeuzagallo commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -553,7 +553,7 @@ static WGPUTextureViewDimension convertViewDimension(WGSL::TextureViewDimension
     if (descriptor.layout) {
         // FIXME: https://bugs.webkit.org/show_bug.cgi?id=257947 - GPUAutoLayout should not rely on the only
```
same

mwyrzykowski committed:
[WebGPU] GPUAutoLayout should have special designation
https://bugs.webkit.org/show_bug.cgi?id=257947
<radar://110624583>

Reviewed by Myles C. Maxfield.

Instead of assuming zero bind group layouts == auto layout, use
an optional to track auto layout.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::m_autoPipelineLayout):
(WebCore::GPUDevice::createAutoPipelineLayout):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createPipelineLayout):
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUPipelineLayoutDescriptor.h:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/PipelineLayout.h:
(WebGPU::PipelineLayout::create):
(WebGPU::PipelineLayout::isValid const):
(WebGPU::PipelineLayout::isAutoLayout const):
* Source/WebGPU/WebGPU/PipelineLayout.mm:
(WebGPU::Device::createPipelineLayout):
(WebGPU::PipelineLayout::PipelineLayout):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPUPipelineLayoutDescriptor.h:

Canonical link: https://commits.webkit.org/265134@main
